vmf/scripts/mods/vmf/modules/vmf_package_manager.lua

```diff
@@ -13,166 +71,139 @@ local _loaded_packages = {}
 --[[
   Loads a mod package.
   * package_name [string]  : package name. needs to be the full path to the `.package` file without the extension
-  * callback     [function]: (optional) callback for asynchronous loading
+  * callback     [function]: (optional) callback for when loading is done
   * sync         [boolean] : (optional) load the packages synchronously, freezing the game until it is loaded
 --]]
 function VMFMod:load_package(package_name, callback, sync)
-  if vmf.check_wrong_argument_type(self, "load_package", "package_name", package_name, "string") or
+  if check_vt1(self, "load_package") or
+     vmf.check_wrong_argument_type(self, "load_package", "package_name", package_name, "string") or
      vmf.check_wrong_argument_type(self, "load_package", "callback", callback, "function", "nil") or
      vmf.check_wrong_argument_type(self, "load_package", "sync", sync, "boolean", "nil")
   then
     return
   end
 
-  if self:has_package_loaded(package_name) then
-    self:error("Package '%s' has already been loaded", package_name)
+  if self:package_status(package_name) == "loaded" then
+    self:error(ERRORS.REGULAR.package_already_loaded, package_name)
     return
   end
 
-  if not _loaded_packages[self] then
-    _loaded_packages[self] = {}
-  end
-
   local resource_package = Mod.resource_package(self:get_internal_data("mod_handle"), package_name)
   if not resource_package then
-    self:error("Could not find package '%s'.", package_name)
+    self:error(ERRORS.REGULAR.package_not_found, package_name)
     return
   end
 
-  local is_loading = self:is_package_loading(package_name)
+  -- If package is_already_queued it means it was already loaded asynchronously before, but not fully loaded yet.
+  -- It can have "queued" or "loading" status. Don't redefine data for this package.
+  local is_already_queued = _packages[package_name] ~= nil
+  if not is_already_queued then
+    _packages[package_name] = {
+      status           = "queued",
+      resource_package = resource_package,
+      callback         = callback,
+      mod              = self
+    }
+  end
 
   if sync then
-    if not is_loading then
+    -- Load resource package if it's not already loading.
+    if _packages[package_name].status == "queued" then
       resource_package:load()
     end
-
-    resource_package:flush()
-
-    _loaded_packages[self][package_name] = resource_package
+    if is_already_queued then
+      remove_package_from_queue(package_name)
+    end
+    flush_package(package_name)
```

Comment:
Yeah, this thing is very questionable for me as well. I didn't want to add it in the beginning and added only because SirAiedail did. I could think only of 1 reasonable use-case: when a lot of mods do async loading on some event, like entering `StateIngame`, and then the game is about to start, but some mod packages are not loaded yet, so the mod can force-load it. This way, when there are not many mods, the game won't freeze. And when there are a lot of mods, the game will freeze only to load remaining packages and won't crash.

Reply:
Another use case would be if multiple systems within a single mod require the same package:

```lua
-- system1.lua
mod:load_package("my_package", function()
  -- use package data
end)

-- system2.lua
mod:load_package("my_package", nil, true)
-- needs package data immediately
```

If `system2` loads the package first, `system1` will receive the "package already loaded" error, which is fine. But if `system1` runs first and `system2` shortly after, before the package is fully loaded, the package manager needs to either switch to forced synchronous loading or return an error.

While an error would be fine from VMF's POV, it would make things complicated for the user. They already need synchronous loading at that point, so they probably can't just try again a few seconds later.

I do agree that it is an issue that only one callback would be called. It was fine with my version, since there were no callbacks for synchronous loading.